clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -951,6 +953,47 @@ raw2trace_t::process_next_thread_buffer(raw2trace_thread_data_t *tdata,
         // Make a copy to avoid clobbering the entry we pass to process_offline_entry()
         // when it calls get_next_entry() on its own.
         offline_entry_t entry = *in_entry;
+        if (tdata->to_inject_syscall_ != -1) {
+            bool is_marker = entry.extended.type == OFFLINE_TYPE_EXTENDED &&
+                entry.extended.ext == OFFLINE_EXT_TYPE_MARKER;
+            bool is_injection_point = false;
+            // We inject the syscall trace after all markers added in the
+            // pre-syscall event.
+            if (
+                // For syscalls not specified in -record_syscall, which do not have
+                // the func_id-func_retval markers.
+                entry.timestamp.type == OFFLINE_TYPE_TIMESTAMP ||
+                // For syscalls that did not have a post-event because the trace ended.
+                (entry.extended.type == OFFLINE_TYPE_EXTENDED &&
+                 entry.extended.ext == OFFLINE_EXT_TYPE_FOOTER) ||
+                // For syscalls interrupted by a signal and did not have a post-syscall
+                // event.
+                (is_marker && entry.extended.valueB == TRACE_MARKER_TYPE_KERNEL_EVENT)) {
+                is_injection_point = true;
+            } else if (is_marker && entry.extended.valueB == TRACE_MARKER_TYPE_FUNC_ID) {
+                if (!tdata->saw_first_func_id_marker_after_syscall_) {
+                    // XXX i#7482: If we allow recording zero args for syscalls in
+                    // -record_syscall, we would need to update this logic.
+                    tdata->saw_first_func_id_marker_after_syscall_ = true;
+                } else {
+                    // For syscalls specified in -record_syscall, for which we inject
+                    // just before the func_id-func_retval markers.
+                    is_injection_point = true;
+                }
+            }
```

Comment:
Turn the above (and maybe below too?) into a helper function to keep the containing function from ballooning.